scvi/external/scar/_model.py

```diff
@@ -0,0 +1,310 @@
+            library_log_means, library_log_vars = _init_library_size(
+                self.adata_manager, n_batch
+            )
+
+        # self.summary_stats provides information about anndata dimensions and other tensor info
+        if not torch.is_tensor(ambient_profile):
+            if isinstance(ambient_profile, str):
+                ambient_profile = np.nan_to_num(adata.varm[ambient_profile])
+            elif isinstance(ambient_profile, pd.DataFrame):
+                ambient_profile = ambient_profile.fillna(0).values
+            elif isinstance(ambient_profile, np.ndarray):
+                ambient_profile = np.nan_to_num(ambient_profile)
+            elif not ambient_profile:
+                ambient_profile = adata.X.sum(axis=0) / adata.X.sum(axis=0).sum()
+                ambient_profile = np.nan_to_num(ambient_profile)
+            else:
+                raise TypeError(
+                    f"Expecting str / np.array / None / pd.DataFrame, but get a {type(ambient_profile)}"
+                )
+            ambient_profile = (
+                torch.from_numpy(np.asarray(ambient_profile)).float().reshape(1, -1)
+            )
+
+        self.module = SCAR_VAE(
+            ambient_profile=ambient_profile,
+            n_input=self.summary_stats.n_vars,
+            n_batch=n_batch,
+            n_labels=self.summary_stats.n_labels,
+            n_continuous_cov=self.summary_stats.get("n_extra_continuous_covs", 0),
+            n_cats_per_cov=n_cats_per_cov,
+            n_hidden=n_hidden,
+            n_latent=n_latent,
+            n_layers=n_layers,
+            dropout_rate=dropout_rate,
+            dispersion=dispersion,
+            gene_likelihood=gene_likelihood,
+            latent_distribution=latent_distribution,
+            use_size_factor_key=use_size_factor_key,
+            library_log_means=library_log_means,
+            library_log_vars=library_log_vars,
+            scale_activation=scale_activation,
+            sparsity=sparsity,
+            **model_kwargs,
+        )
+        self._model_summary_string = (
+            "SCVI-AR Model with the following params: \nn_hidden: {}, n_latent: {}, n_layers: {}, dropout_rate: "
+            "{}, dispersion: {}, gene_likelihood: {}, latent_distribution: {}"
+        ).format(
+            n_hidden,
+            n_latent,
+            n_layers,
+            dropout_rate,
+            dispersion,
+            gene_likelihood,
+            latent_distribution,
+        )
+        self.init_params_ = self._get_init_params(locals())
+
+    @classmethod
+    @setup_anndata_dsp.dedent
+    def setup_anndata(
+        cls,
+        adata: AnnData,
+        layer: Optional[str] = None,
+        batch_key: Optional[str] = None,
+        labels_key: Optional[str] = None,
+        size_factor_key: Optional[str] = None,
+        categorical_covariate_keys: Optional[List[str]] = None,
+        continuous_covariate_keys: Optional[List[str]] = None,
```

Comment:
I haven't read scAR too in depth -- does it make sense to allow this functionality? My impression is that it's run once per 10x lane?

Reply:
Therefore, no batch correction, etc is necessary. If this is the case, can you remove this as well as all related functionality at the PyTorch module level?

Reply:
> I haven't read scAR too in depth -- does it make sense to allow this functionality? My impression is that it's run once per 10x lane?

Hi @adamgayoso , I am the developer of scAR. It's run once per experiment/batch -- the same experiment and batch should have the same ambient profile (a.k.a same composition of contamination), it is lane-independent.

Reply:
Thanks @CaibinSh, so it's fair to say that there is no notion of batch correction with this model so we should ten remove all these fields, except perhaps the size factor.

Reply:
I removed all of the keys and fields related to batch, label and covariates in the latest commit. I did leave them in the `setup_anndata()` function though in order to use the `generic_forward()` and other useful scVI utility functions that would break otherwise. That should be fine right? Alternatively we could rewrite these helper functions in the model to not rely on batch and label keys